internal/nvmeof/controller/controllerserver.go

```diff
@@ -330,16 +337,134 @@ func (cs *Server) ControllerModifyVolume(
 	return &csi.ControllerModifyVolumeResponse{}, nil
 }
 
-// ControllerExpandVolume handles volume expansion requests.
-// For now it only updates the capacity in the response as NVMe-oF
-// this must be added because ControllerModifyVolume requires the sidecar csi-resizer. and
-// csi-resizer searches for the capacity ControllerServiceCapability_RPC_EXPAND_VOLUME.
-// In the future, if NVMe-oF gateway supports volume expansion, the logic must be added here.
 func (cs *Server) ControllerExpandVolume(
 	ctx context.Context,
 	req *csi.ControllerExpandVolumeRequest,
 ) (*csi.ControllerExpandVolumeResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "ControllerExpandVolume is not implemented for NVMe-oF volumes")
+	err := cs.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_EXPAND_VOLUME)
+	if err != nil {
+		log.ErrorLog(ctx, "invalid expand volume req: %v", err)
+
+		return nil, err
+	}
+
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "volume ID cannot be empty")
+	}
+
+	capRange := req.GetCapacityRange()
+	if capRange == nil {
+		return nil, status.Error(codes.InvalidArgument, "capacityRange cannot be empty")
+	}
+	requestedBytes := capRange.GetRequiredBytes()
+	// Validate it's positive (as per CSI spec)
+	if requestedBytes <= 0 {
+		return nil, status.Error(codes.InvalidArgument, "requested capacity must be positive")
+	}
+	// Round to MiB/GiB boundary (returns bytes)
+	roundedBytes := util.RoundOffBytes(requestedBytes)           // int64, in bytes
+	requestedMiB := uint64(util.RoundOffVolSize(requestedBytes)) // Convert to MiB for gateway
+
+	// Lock volume
+	if acquired := cs.volumeLocks.TryAcquire(volumeID); !acquired {
+		log.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, volumeID)
+
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+	defer cs.volumeLocks.Release(volumeID)
+
+	secrets := req.GetSecrets()
+	if secrets == nil {
+		/*	Reuse the existing function - it's just getting Ceph credentials from k8s secret.
+			GetControllerPublishSecretRef name is misleading here.
+			another option is to create a new function like this one:
+
+				func GetControllerSecretRef(volumeId, driverType string) (string, string, error) {
+					return GetControllerPublishSecretRef(volumeId, driverType)
+				}
+		*/
+		secretName, secretNamespace, err := util.GetControllerPublishSecretRef(req.GetVolumeId(), util.RBDType)
```

Comment:
The `ControllerExpandVolumeRequest` should contain secrets already, use those.

https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume

Reply:
you may remove the commented out code :smile: